File: 284.py
import string

mod = 14
N = 100
ans = 0

def rep(x):
    s = ""
    sum = 0
    while x > 0:
        s = string.hexdigits[x % mod] + s
        sum += x % mod
        x /= mod
    return s, sum

Q = []
for i in range(1, mod):
    if i * i % mod == i:
        Q += [i]
        ans += i
        print rep(i)[0]

for n in range(1, N):
    print n
    mpow = mod**n
    P = []
    for x in Q:
        y = x
        for i in range(mod):
            if y * y / mpow % mod == i:
                P += [y]
                if i != 0:
                    s, sum = rep(y)
                    ans += sum
                    # print s
            y += mpow
    Q = P

print 'ans = %d' % (ans)
print rep(604557993)

File: 284.cpp
#include <cstdio>
using namespace std;

const int mod = 14, n = 10000, N = n * 2 + 10;

int x[N];

int main() {
  x[0] = 49;
  long long ans = 1;
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j <= i * 2; ++j) {
      x[j + 1] += x[j] / mod;
      x[j] %= mod;
    }

    int sx = 0, sy = 2;
    for (int j = 0; j <= i; ++j) {
      sx += x[j];
      sy += mod - 1 - x[j];
    }
    if (x[i] != 0)
      ans += sx;
    if (x[i] != mod - 1)
      ans += sy;

    int t = x[i + 1]; // the new digit
    printf("%c", (t < 10 ? t + '0' : t + 'a' - 10));
    for (int j = 0; j <= i; ++j) {
      x[i + 1 + j] += 2 * t * x[j];
    }
    x[i * 2 + 2] += t * t;
  }
  printf("\nans = %lld\n", ans);
  return 0;
}
